### PR TITLE
[WIP] Improve trigger character handling

### PR DESCRIPTION
**Pull request by hrsh7th**
Fixes https://github.com/prabirshrestha/asyncomplete.vim/issues/173

New implementation detect below cases as text changed.

### InsertEnter
When `InsertEnter` to the tag attributes, I hope to show attribute's candidates.

```html
<div aria-autocomplete="|"></div>
```

### Line changes.
When apply `additionalTextEdits`, I hope to show imported package's candidates.

```go
package main

import "fmt"

func main() {
  fmt.|
}
```

### CODE REVIEW

File: autoload/asyncomplete/utils/_on_change/timer.vim [hunk 1]
```diff
@@ -30,22 +30,25 @@ function! s:unregister(obj, cb) abort
 endfunction
 
 function! s:on_insert_enter() abort
-    let s:previous_position = getcurpos()
-    call s:change_tick_start()
+    call s:change_tick_start(v:true)
```
Comment (hrsh7th) [inline autoload/asyncomplete/utils/_on_change/timer.vim]: Force detect text changes on InsertEnter.

**Comment by prabirshrestha** [comment]
what happens if you press enter and go to next line? does it open the popup?

I can't remember exactly why I added the cursor line check but it was so that it doesn't randomly open the pum.

I will try this PR in the mean time for a day or say before I merge it in.

**Comment by hrsh7th** [comment]
I checked <CR>.
Unfortunately, unwanted pum was displayed when using asyncomplete-lsp.vim.

I think the problem should fix in asyncomplete-lsp side bacause probably the below case want to detect text changed as line changes.

For example, When the example language server returns `,` as trigger character.

```typescript
type ExampleObject = {
  SomeProp1: "some1-1" | "some1-2";
  SomeProp2: "some2-1" | "some2-2"
};

const exampleObject: ExampleObject = {
  SomeProp1: "some1-1",|<CR>
};
```

`typescript-language-server` can return remaining property of `ExampleObject` that is `SomeProp2`.

**Comment by prabirshrestha** [comment]
Even I'm not sure what is the best fix because it is easy to open the popup by calling force refresh.

```vim
imap <c-space> <Plug>(asyncomplete_force_refresh)
```

Another option is to fire an event from `vim-lsp` so `asyncomplete-lsp` can register `LspTextEditApplyPre` and `LspTextEdit`. In `LspTextEditApplyPre` you register the state of pum as well as the trigger commit and `LspTextEdit` call `asyncomplete_force_refresh` if the pum should open or should remain closed.

**Comment by hrsh7th** [comment]
I wondering why asyncomplete shows candidates when press `<CR>`.

When the case, cursor line is `    |` so I don't think to match the `\k\+`.
My guessing is this behavior occurs asyncomplete's cache logic...maybe.

I will more deep investigate to solve it.

Although, your advice implementation is very easy and has no possibility of side-effect in my think.
I will try to implement it (with found the time...)

**Comment by prabirshrestha** [comment]
I think it is primarily because of how `TextChangedI` and `TextChangedP` gets triggered. I can't remember the details now but I had tried to find multiple solutions and the best I could find was the line number. 

Even I'm not happy with the current solution so if you find a better one feel free to share.

**Comment by hrsh7th** [comment]
I found the reason of asyncomplete showing pum when press `<CR>` with this patch.

1. Press `<CR>` and then trigger `text changed`.
2. Call `s:on_change` in `asyncomplete.vim`.
3. Run this `if-statement` https://github.com/prabirshrestha/asyncomplete.vim/pull/174/commits/e2d3cbf176bc1e517935b075d26c1e5f2ab180bf#diff-4948f573df0018e59dd56d5fcdcf473eL290
4. New line has no characters so `l:startcol` is always -1 (no mark to `l:sources_to_notify`)
5. Call `s:trigger` and `s:update_pum` then the cached matches showing to the pum.

I think we should modify below changes... maybe.

#### 1. Change trigger_character checking approach.
- Currently, asyncomplete searches trigger_character from col - 1 only.
- I think we should check trigger_character like `lsp#utils#_get_before_char_skip_white`.

#### 2. If the lnum was changed, clear cached source results
- It should only do clear cache, we don't change the approach adding `l:sources_to_notify`.

It hard to explain so I try to implment it.

**Comment by hrsh7th** [comment]
I've updated this PR to improve trigger character handling.

**Comment by hrsh7th** [comment]
This PR already works but I should implement trigger character caching.

File: autoload/asyncomplete.vim [hunk 2]
```diff
@@ -421,6 +426,10 @@ function! s:recompute_pum(...) abort
 
         if l:startcol > l:ctx['col']
             call asyncomplete#log('core', 's:recompute_pum', 'ignoring due to wrong start col', l:startcol, l:ctx['col'])
+            echomsg string(["'invalid cols'", {
```
Comment (prabirshrestha) [inline autoload/asyncomplete.vim]: any reason why `echomsg` instead of just `asyncomplete#log`?

Comment (hrsh7th) [reply autoload/asyncomplete.vim]: Sorry... It is wrong commit! I'll revert it.

**Comment by prabirshrestha** [comment]
@hrsh7th  Is this ready for review?

**Comment by hrsh7th** [comment]
Sorry for my delay... I close this PR.
The problem was solved by https://github.com/prabirshrestha/asyncomplete.vim/pull/186, I think.

**Pull request closed without merging**